storage: Add reservation system for stores.

BramGruneir opened the pull request:
This change adds a reservation system (a bookie). Note that this does not yet
add it to the store.

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/6820)

<!-- Reviewable:end -->

BramGruneir committed:
storage: Add reservation system for stores.

This change adds a reservation system (a bookie). Note that this does not yet
add it to the store.

tschottdorf commented on storage/reservation.go:
```diff
@@ -0,0 +1,247 @@
+}
+
+// fillBookingLocked fills a booking. It requires that the reservation lock is
+// held. This should only be called internally.
```
consider exporting the methods which you expect to be called externally (they won't show up in godoc since the receiver isn't exported).

tschottdorf commented on storage/reservation.go:
```diff
@@ -0,0 +1,247 @@
+	clock              *hlc.Clock
+	reservationTimeout time.Duration // How long each reservation is held.
+	mu                 struct {
+		sync.RWMutex                              // Protects all values within the mu struct.
```
why not a regular mutex?

tschottdorf commented on storage/reservation.go:
```diff
@@ -0,0 +1,247 @@
+	return b.mu.size
+}
+
+// reserve resreves a new replica.
```
s/srev/serv/

cuongdo commented:
Review status: 0 of 2 files reviewed at latest revision, 11 unresolved discussions, all commit checks successful.

---

*[storage/reservation.go, line 45 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6820#-KIDSoKTIQH-IH4zUeUo:-KIDSoKTIQH-IH4zUeUp:-2121357320) ([raw file](https://github.com/cockroachdb/cockroach/blob/800be2c7b142625528b383f255fc9782dfb4dd7c/storage/reservation.go#L45)):*
> ```Go
> 	expireAt    roachpb.Timestamp // This is also the priority for the queue.
> 	filled      bool              // Has this booking been filled?
> 	index       int               // index of the item in the heap, required for heap.Interface
> ```

How is `index` required by `heap.Interface`? I think it's more accurate to say that it's required to maintain the priority queue and is updated by the `heap.Interface` methods.

---

*[storage/reservation.go, line 48 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6820#-KIDURZCZj8OxQqLdVoR:-KIDURZDtZOi565qmgPs:-1010715928) ([raw file](https://github.com/cockroachdb/cockroach/blob/800be2c7b142625528b383f255fc9782dfb4dd7c/storage/reservation.go#L48)):*
> ```Go
> }
> 
> // bookingPQ implements the heap.Interface (which includes sort.Interface) and
> ```

It would be clearer if you added something about how this heap is organized. such as:

* is this a min/max heap?
* what is it minimizing or maximizing (`expireAt`)?

After quickly reviewing this comment, I should know these basic things about the heap's behavior rather than having to look at the `heap.Interface` methods below.

---

*[storage/reservation.go, line 91 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6820#-KIDW6pm_F13jlG5U8et:-KIDW6pm_F13jlG5U8eu:-1788243512) ([raw file](https://github.com/cockroachdb/cockroach/blob/800be2c7b142625528b383f255fc9782dfb4dd7c/storage/reservation.go#L91)):*
> ```Go
> 		return nil
> 	}
> 	return (pq)[0]
> ```

Why does `pq` have parentheses around it?

---

*[storage/reservation.go, line 94 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6820#-KIDXXrXPHJrLXpIKc5t:-KIDXXrXPHJrLXpIKc5u:-276399430) ([raw file](https://github.com/cockroachdb/cockroach/blob/800be2c7b142625528b383f255fc9782dfb4dd7c/storage/reservation.go#L94)):*
> ```Go
> }
> 
> // enqueue either adds the booking to the queue or updates its location in the
> ```

What is the intended use case for updating a booking's location in the queue?

---

*[storage/reservation.go, line 96 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6820#-KIDU0tr7stBJJbwq6JA:-KIDU0tr7stBJJbwq6JB:1019410749) ([raw file](https://github.com/cockroachdb/cockroach/blob/800be2c7b142625528b383f255fc9782dfb4dd7c/storage/reservation.go#L96)):*
> ```Go
> // enqueue either adds the booking to the queue or updates its location in the
> // priority queue.
> func (pq *bookingPQ) enqueue(resTimeout *booking) {
> ```

why is this parameter called `resTimeout`?

---

*[storage/reservation.go, line 137 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6820#-KIDYqS4YKaSctwRAAJT:-KIDYqS4YKaSctwRAAJU:-927580178) ([raw file](https://github.com/cockroachdb/cockroach/blob/800be2c7b142625528b383f255fc9782dfb4dd7c/storage/reservation.go#L137)):*
> ```Go
> 
> // len returns the total number of outstanding reservations.
> func (b *bookie) len() int {
> ```

Maybe change `len` to a more obvious name? Like `outstanding`?

---

*[storage/reservation.go, line 144 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6820#-KIDZ2SZMc-hmxnqIjie:-KIDZ2SZMc-hmxnqIjif:-298280763) ([raw file](https://github.com/cockroachdb/cockroach/blob/800be2c7b142625528b383f255fc9782dfb4dd7c/storage/reservation.go#L144)):*
> ```Go
> 
> // size returns the total bytes reserved.
> func (b *bookie) size() int64 {
> ```

`size` isn't obvious either, especially since you also have a `len` method. Consider changing one or both.

---

*[storage/reservation.go, line 150 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6820#-KIDZAM0BlLYdLK27qBj:-KIDZAM0BlLYdLK27qBk:-1709532453) ([raw file](https://github.com/cockroachdb/cockroach/blob/800be2c7b142625528b383f255fc9782dfb4dd7c/storage/reservation.go#L150)):*
> ```Go
> }
> 
> // reserve resreves a new replica.
> ```

s/resreves/reserves/

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6820)*
<!-- Sent from Reviewable.io -->

cuongdo commented:
Review status: 0 of 2 files reviewed at latest revision, 18 unresolved discussions, all commit checks successful.

---

*[storage/reservation_test.go, line 39 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6820#-KIDcDWhcDVWbK0qnvo5:-KIDcDWhcDVWbK0qnvo6:-1108467673) ([raw file](https://github.com/cockroachdb/cockroach/blob/800be2c7b142625528b383f255fc9782dfb4dd7c/storage/reservation_test.go#L39)):*
> ```Go
> 
> // TestReserve ensures that you can never have more than one reservation for a
> // perticual rangeID at a time, and that both `reserve` and `fill` function
> ```

`peticual` -> `particular`?

---

*[storage/reservation_test.go, line 48 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6820#-KIDcpq2Iv8ShZj6EJM2:-KIDcpq2Iv8ShZj6EJM3:-1938148189) ([raw file](https://github.com/cockroachdb/cockroach/blob/800be2c7b142625528b383f255fc9782dfb4dd7c/storage/reservation_test.go#L48)):*
> ```Go
> 	testCases := []struct {
> 		rangeID          int
> 		reserve          bool // true for reserve, false for fill
> ```

it may make the code more obvious to change reserve to something like `type action int` then use constants to indicate whether you want to reserve or fill.

---

*[storage/reservation_test.go, line 49 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6820#-KIDdmVA5l2Ly1wBHOqR:-KIDdmVA5l2Ly1wBHOqS:-1458411345) ([raw file](https://github.com/cockroachdb/cockroach/blob/800be2c7b142625528b383f255fc9782dfb4dd7c/storage/reservation_test.go#L49)):*
> ```Go
> 		rangeID          int
> 		reserve          bool // true for reserve, false for fill
> 		expected         bool
> ```

maybe make this `expSuccess`? it's not immediately clear what `testCase.expected` means just by reading that.

---

*[storage/reservation_test.go, line 54 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6820#-KIDdIx2E6E4pf2ZCr1J:-KIDdIx2E6E4pf2ZCr1K:107586298) ([raw file](https://github.com/cockroachdb/cockroach/blob/800be2c7b142625528b383f255fc9782dfb4dd7c/storage/reservation_test.go#L54)):*
> ```Go
> 		expectedTimeouts int
> 	}{
> 		{1, true, true, 1, 1, 1},
> ```

This list of testcases is hard to read without field names. My eyes have to bounce back and forth between the struct declaration and this slice to comprehend things. Consider adding them to make the testcases easier to read. Maybe you can abbreviate "expected" as "exp" to make each test case fit on a single line.

---

*[storage/reservation_test.go, line 145 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6820#-KIDecvShEhIT1hYH8g7:-KIDecvT0VR0ZCifX1Tm:354799840) ([raw file](https://github.com/cockroachdb/cockroach/blob/800be2c7b142625528b383f255fc9782dfb4dd7c/storage/reservation_test.go#L145)):*
> ```Go
> // reserved 100 bytes.
> func verifyBookie(t *testing.T, b *bookie, reservations, queueLen int) {
> 	if e, a := int64(reservations*100), b.size(); e != a {
> ```

You should make `100` a constant, and use it in `TestReservationQueue` so that the intent is more obvious and to avoid trivial errors. Or. pass in the reservation size as a parameter.

---

*[storage/reservation_test.go, line 167 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6820#-KIDesy4Mod3EAz02pRo:-KIDesy4Mod3EAz02pRp:-1365593720) ([raw file](https://github.com/cockroachdb/cockroach/blob/800be2c7b142625528b383f255fc9782dfb4dd7c/storage/reservation_test.go#L167)):*
> ```Go
> 	// Load a collection of reservations into the bookie.
> 	for i := 1; i <= 10; i++ {
> 		// Ensure all the timesouts are 10 nanoseconds apart.
> ```

s/timesouts/timeouts/

Also. should the comment say 10 ns or 100 ns? (I ask because you're incrementing `mc` by 100.)

---

*[storage/reservation_test.go, line 178 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6820#-KIDfE80fXu0MCF1r6xU:-KIDfE80fXu0MCF1r6xV:40154954) ([raw file](https://github.com/cockroachdb/cockroach/blob/800be2c7b142625528b383f255fc9782dfb4dd7c/storage/reservation_test.go#L178)):*
> ```Go
> 		}
> 	}
> 	verifyBookie(t, b, 10, 10)
> ```

consider adding inline comments (e.g. `/* reservations */`) to enhance readability of these calls

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6820)*
<!-- Sent from Reviewable.io -->

BramGruneir commented:
Sorry about the spelling errors in comments.  I just switched to a new editor and it's lacking spellchecking in comments.  I'm going to see if I can do something about that.

<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary>
> #### storage: Add reservation system for stores\.
> 
> This change adds a reservation system (a bookie). Note that this does not yet
> 
> add it to the store.
> 
> 
> 
</details>


---

Review status: 0 of 2 files reviewed at latest revision, 18 unresolved discussions, some commit checks pending.

---

*[storage/reservation.go, line 45 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6820#-KIDSoKTIQH-IH4zUeUo:-KIGqrcmYqCBciGWVa6Q:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/800be2c7b142625528b383f255fc9782dfb4dd7c/storage/reservation.go#L45)):*
<details><summary><i>Previously, cuongdo (Cuong Do) wrote…</i></summary>
> How is `index` required by `heap.Interface`? I think it's more accurate to say that it's required to maintain the priority queue and is updated by the `heap.Interface` methods.
</details>
Done.

---

*[storage/reservation.go, line 48 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6820#-KIDURZCZj8OxQqLdVoR:-KIGdVDY85k0G3S-yI3v:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/800be2c7b142625528b383f255fc9782dfb4dd7c/storage/reservation.go#L48)):*
<details><summary><i>Previously, cuongdo (Cuong Do) wrote…</i></summary>
> It would be clearer if you added something about how this heap is organized. such as:
> 
> * is this a min/max heap?
> * what is it minimizing or maximizing (`expireAt`)?
> 
> After quickly reviewing this comment, I should know these basic things about the heap's behavior rather than having to look at the `heap.Interface` methods below.
</details>
Done.

---

*[storage/reservation.go, line 91 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6820#-KIDW6pm_F13jlG5U8et:-KIGc3KSqJnbMpF9yLHs:1373558590) ([raw file](https://github.com/cockroachdb/cockroach/blob/800be2c7b142625528b383f255fc9782dfb4dd7c/storage/reservation.go#L91)):*
<details><summary><i>Previously, cuongdo (Cuong Do) wrote…</i></summary>
> Why does `pq` have parentheses around it?
</details>
no need.

---

*[storage/reservation.go, line 94 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6820#-KIDXXrXPHJrLXpIKc5t:-KIGR9vjmJq08YUILFBX:-1984073786) ([raw file](https://github.com/cockroachdb/cockroach/blob/800be2c7b142625528b383f255fc9782dfb4dd7c/storage/reservation.go#L94)):*
<details><summary><i>Previously, cuongdo (Cuong Do) wrote…</i></summary>
> What is the intended use case for updating a booking's location in the queue?
</details>
 These were useful for the store pool (which I cribbed this priority queue from), that requires frequent updates to values within the queue.  I've removed it.

---

*[storage/reservation.go, line 96 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6820#-KIDU0tr7stBJJbwq6JA:-KIGMp_yD8JqKlsuEX2d:2028613047) ([raw file](https://github.com/cockroachdb/cockroach/blob/800be2c7b142625528b383f255fc9782dfb4dd7c/storage/reservation.go#L96)):*
<details><summary><i>Previously, cuongdo (Cuong Do) wrote…</i></summary>
> why is this parameter called `resTimeout`?
</details>
Old name, fixed.

---

*[storage/reservation.go, line 117 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6820#-KIBRi8iNrPjUDWDVQ5k-r1-117:-KIGc6dqA4v8FtAfI46o:-1773286888) ([raw file](https://github.com/cockroachdb/cockroach/blob/800be2c7b142625528b383f255fc9782dfb4dd7c/storage/reservation.go#L117)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> why not a regular mutex?
</details>
So we can have multiple reads occurring at once, if there are multiple incoming reservation requests.  It's a minor optimization, but since the the locking of the bookie is fairly simplistic, I see no reason not to use the RW.  Is there a reason not to use it?

---

*[storage/reservation.go, line 137 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6820#-KIDYqS4YKaSctwRAAJT:-KIGcg-sXctT_rjA69hC:-84779010) ([raw file](https://github.com/cockroachdb/cockroach/blob/800be2c7b142625528b383f255fc9782dfb4dd7c/storage/reservation.go#L137)):*
<details><summary><i>Previously, cuongdo (Cuong Do) wrote…</i></summary>
> Maybe change `len` to a more obvious name? Like `outstanding`?
</details>
Sure, done.

---

*[storage/reservation.go, line 144 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6820#-KIDZ2SZMc-hmxnqIjie:-KIGcwVmEpC89r6GUobb:1738287543) ([raw file](https://github.com/cockroachdb/cockroach/blob/800be2c7b142625528b383f255fc9782dfb4dd7c/storage/reservation.go#L144)):*
<details><summary><i>Previously, cuongdo (Cuong Do) wrote…</i></summary>
> `size` isn't obvious either, especially since you also have a `len` method. Consider changing one or both.
</details>
renamed to `ReservedBytes`

---

*[storage/reservation.go, line 150 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6820#-KIBRi8iNrPjUDWDVQ5k-r1-150:-KIGdKzgFwjjptWeRUuT:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/800be2c7b142625528b383f255fc9782dfb4dd7c/storage/reservation.go#L150)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> s/srev/serv/
</details>
Done.

---

*[storage/reservation.go, line 150 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6820#-KIDZAM0BlLYdLK27qBj:-KIGdLWVOrM_judypuNM:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/800be2c7b142625528b383f255fc9782dfb4dd7c/storage/reservation.go#L150)):*
<details><summary><i>Previously, cuongdo (Cuong Do) wrote…</i></summary>
> s/resreves/reserves/
</details>
Done.

---

*[storage/reservation.go, line 192 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6820#-KIBRi8iNrPjUDWDVQ5k-r1-192:-KIGdZnD8tfhtX5k_oT_:124292462) ([raw file](https://github.com/cockroachdb/cockroach/blob/800be2c7b142625528b383f255fc9782dfb4dd7c/storage/reservation.go#L192)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> consider exporting the methods which you expect to be called externally (they won't show up in godoc since the receiver isn't exported).
</details>
Done. I like this idea.

---

*[storage/reservation_test.go, line 39 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6820#-KIDcDWhcDVWbK0qnvo5:-KIGeCIqB7HS9wffFdYH:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/800be2c7b142625528b383f255fc9782dfb4dd7c/storage/reservation_test.go#L39)):*
<details><summary><i>Previously, cuongdo (Cuong Do) wrote…</i></summary>
> `peticual` -> `particular`?
</details>
Done.

---

*[storage/reservation_test.go, line 48 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6820#-KIDcpq2Iv8ShZj6EJM2:-KIGpDUXIEfIE2uRD-FU:1203469616) ([raw file](https://github.com/cockroachdb/cockroach/blob/800be2c7b142625528b383f255fc9782dfb4dd7c/storage/reservation_test.go#L48)):*
<details><summary><i>Previously, cuongdo (Cuong Do) wrote…</i></summary>
> it may make the code more obvious to change reserve to something like `type action int` then use constants to indicate whether you want to reserve or fill.
</details>
That seems like a bit much, there are only two actions.  

---

*[storage/reservation_test.go, line 49 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6820#-KIDdmVA5l2Ly1wBHOqR:-KIGeSAOJPOvUVjuF4d1:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/800be2c7b142625528b383f255fc9782dfb4dd7c/storage/reservation_test.go#L49)):*
<details><summary><i>Previously, cuongdo (Cuong Do) wrote…</i></summary>
> maybe make this `expSuccess`? it's not immediately clear what `testCase.expected` means just by reading that.
</details>
Done.

---

*[storage/reservation_test.go, line 54 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6820#-KIDdIx2E6E4pf2ZCr1J:-KIGeXW7keGctS4QN7Kg:1707216754) ([raw file](https://github.com/cockroachdb/cockroach/blob/800be2c7b142625528b383f255fc9782dfb4dd7c/storage/reservation_test.go#L54)):*
<details><summary><i>Previously, cuongdo (Cuong Do) wrote…</i></summary>
> This list of testcases is hard to read without field names. My eyes have to bounce back and forth between the struct declaration and this slice to comprehend things. Consider adding them to make the testcases easier to read. Maybe you can abbreviate "expected" as "exp" to make each test case fit on a single line.
</details>
Ha, I've been told to remove the field names in the past due to verbosity.  Done.

---

*[storage/reservation_test.go, line 145 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6820#-KIDecvShEhIT1hYH8g7:-KIGnCixDpMmwlcQZJXq:661510352) ([raw file](https://github.com/cockroachdb/cockroach/blob/800be2c7b142625528b383f255fc9782dfb4dd7c/storage/reservation_test.go#L145)):*
<details><summary><i>Previously, cuongdo (Cuong Do) wrote…</i></summary>
> You should make `100` a constant, and use it in `TestReservationQueue` so that the intent is more obvious and to avoid trivial errors. Or. pass in the reservation size as a parameter.
</details>
Passed as a parameter and I moved verifyBookie up so it could be used in both tests.

---

*[storage/reservation_test.go, line 167 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6820#-KIDesy4Mod3EAz02pRo:-KIGpO4qGA1i3061yAnH:-1705807047) ([raw file](https://github.com/cockroachdb/cockroach/blob/800be2c7b142625528b383f255fc9782dfb4dd7c/storage/reservation_test.go#L167)):*
<details><summary><i>Previously, cuongdo (Cuong Do) wrote…</i></summary>
> s/timesouts/timeouts/
> 
> Also. should the comment say 10 ns or 100 ns? (I ask because you're incrementing `mc` by 100.)
</details>
Should have been 100. Fixed.

---

*[storage/reservation_test.go, line 178 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6820#-KIDfE80fXu0MCF1r6xU:-KIGnuKeFXWnwVZ1wG5x:-84779010) ([raw file](https://github.com/cockroachdb/cockroach/blob/800be2c7b142625528b383f255fc9782dfb4dd7c/storage/reservation_test.go#L178)):*
<details><summary><i>Previously, cuongdo (Cuong Do) wrote…</i></summary>
> consider adding inline comments (e.g. `/* reservations */`) to enhance readability of these calls
</details>
Sure, done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6820#-:-KIGdc_Q6ctuhfCIJa6Q:8865193)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Review status: 0 of 2 files reviewed at latest revision, 23 unresolved discussions, all commit checks successful.

---

*[storage/reservation.go, line 117 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6820#-KIBRi8iNrPjUDWDVQ5k-r1-117:-KIKCG3TcgJ7E7GdJ_J4:684394116) ([raw file](https://github.com/cockroachdb/cockroach/blob/800be2c7b142625528b383f255fc9782dfb4dd7c/storage/reservation.go#L117)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary>
> So we can have multiple reads occurring at once, if there are multiple incoming reservation requests.  It's a minor optimization, but since the the locking of the bookie is fairly simplistic, I see no reason not to use the RW.  Is there a reason not to use it?
</details>
But `Reserve` gets a write lock. I'm not seeing enough use of the read lock to make this a net win. 

---

*[storage/reservation.go, line 50 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6820#-KIKBNr9OB3NHhxojL3W:-KIKBNrB2Pp1EDuWe-8i:807007553) ([raw file](https://github.com/cockroachdb/cockroach/blob/939f4590198c9a2fafbb0917a7afd53f0ef0181f/storage/reservation.go#L50)):*
> ```Go
> }
> 
> // bookingPQ is a priority queue for bookings (maintained by using a heap)
> ```

A priority queue seems like overkill for this. You'll never have very many active reservations; that's the whole point of the reservation system. You can just loop over the entire set to find expired ones. 

Without a priority queue, it doesn't look like you need the `booking.filled` attribute either (it seems to just be used because you can't easily delete from a priority queue), and then I'd just add `expireAt` to the reservation itself so you don't need two types. 

---

*[storage/reservation.go, line 149 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6820#-KIKDy62j3UF3CYIf8f6:-KIKDy63rZVxcdf85cgK:1610193096) ([raw file](https://github.com/cockroachdb/cockroach/blob/939f4590198c9a2fafbb0917a7afd53f0ef0181f/storage/reservation.go#L149)):*
> ```Go
> }
> 
> // Reserve reserves a new replica.
> ```

s/reserves //

"Reserve a new replica" is a perfectly good sentence; this is the style that golint is trying to encourage by requiring the comment to start with the method name.

What does the return value mean? 

---

*[storage/reservation.go, line 155 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6820#-KIKEVMddZT2i4aBz8HS:-KIKEVMddZT2i4aBz8HT:-1494675801) ([raw file](https://github.com/cockroachdb/cockroach/blob/939f4590198c9a2fafbb0917a7afd53f0ef0181f/storage/reservation.go#L155)):*
> ```Go
> 	if oldRes, ok := b.mu.resByRangeID[res.rangeID]; ok {
> 		if log.V(2) {
> 			log.Infof("there is already an existing reservation for rangeID:%d, %v", res.rangeID,
> ```

When this is exposed as an RPC, a retried RPC should be able to succeed. This should return true if the new reservation matches the old (same range ID and source node/store IDs. We could allow updates to the size if that's useful). 

---

*[storage/reservation.go, line 158 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6820#-KIKEDSljYOFuSZvP7dU:-KIKEDSm4JU1Qyodsfm3:-1827173543) ([raw file](https://github.com/cockroachdb/cockroach/blob/939f4590198c9a2fafbb0917a7afd53f0ef0181f/storage/reservation.go#L158)):*
> ```Go
> 				oldRes)
> 		}
> 		return false
> ```

Doesn't this need to reject reservations when it's too full, not just if it gets a redundant one?

---

*[storage/reservation_test.go, line 29 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6820#-KIKFncYDfQC697U9CYJ:-KIKFncYDfQC697U9CYK:-1904241868) ([raw file](https://github.com/cockroachdb/cockroach/blob/939f4590198c9a2fafbb0917a7afd53f0ef0181f/storage/reservation_test.go#L29)):*
> ```Go
> )
> 
> // createTestBookie creates a new bookie, stopp and manual clock for testing.
> ```

s/pp/pper/

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6820)*
<!-- Sent from Reviewable.io -->

BramGruneir commented:
Review status: 0 of 2 files reviewed at latest revision, 23 unresolved discussions.

---

*[storage/reservation.go, line 117 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6820#-KIBRi8iNrPjUDWDVQ5k-r1-117:-KIVhnmJsdOeKTDeyZjl:761816627) ([raw file](https://github.com/cockroachdb/cockroach/blob/800be2c7b142625528b383f255fc9782dfb4dd7c/storage/reservation.go#L117)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> But `Reserve` gets a write lock. I'm not seeing enough use of the read lock to make this a net win. 
</details>
Changed to a standard lock.

---

*[storage/reservation.go, line 50 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6820#-KIKBNr9OB3NHhxojL3W:-KIWBs-RRANKketMBNme:-1062393254) ([raw file](https://github.com/cockroachdb/cockroach/blob/939f4590198c9a2fafbb0917a7afd53f0ef0181f/storage/reservation.go#L50)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> A priority queue seems like overkill for this. You'll never have very many active reservations; that's the whole point of the reservation system. You can just loop over the entire set to find expired ones. 
> 
> Without a priority queue, it doesn't look like you need the `booking.filled` attribute either (it seems to just be used because you can't easily delete from a priority queue), and then I'd just add `expireAt` to the reservation itself so you don't need two types. 
</details>
I'm going to convert it to a proper fifo queue. 

---

*[storage/reservation.go, line 149 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6820#-KIKDy62j3UF3CYIf8f6:-KIWAqNOVMQHk9zUBtvg:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/939f4590198c9a2fafbb0917a7afd53f0ef0181f/storage/reservation.go#L149)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> s/reserves //
> 
> "Reserve a new replica" is a perfectly good sentence; this is the style that golint is trying to encourage by requiring the comment to start with the method name.
> 
> What does the return value mean? 
</details>
Done.

---

*[storage/reservation.go, line 155 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6820#-KIKEVMddZT2i4aBz8HS:-KIVj-PyYgaasJGKRMt8:549808182) ([raw file](https://github.com/cockroachdb/cockroach/blob/939f4590198c9a2fafbb0917a7afd53f0ef0181f/storage/reservation.go#L155)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> When this is exposed as an RPC, a retried RPC should be able to succeed. This should return true if the new reservation matches the old (same range ID and source node/store IDs. We could allow updates to the size if that's useful). 
</details>
Good point.  I've updated the code.

---

*[storage/reservation.go, line 158 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6820#-KIKEDSljYOFuSZvP7dU:-KIVieB_wfbzq9DhbWfA:-1313942389) ([raw file](https://github.com/cockroachdb/cockroach/blob/939f4590198c9a2fafbb0917a7afd53f0ef0181f/storage/reservation.go#L158)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> Doesn't this need to reject reservations when it's too full, not just if it gets a redundant one?
</details>
I'm planning on adding that soon.  I'm not sure if that should fit within the reservation system or within the store.

---

*[storage/reservation_test.go, line 29 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6820#-KIKFncYDfQC697U9CYJ:-KIVhlUVPB5fpg3-nEJ0:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/939f4590198c9a2fafbb0917a7afd53f0ef0181f/storage/reservation_test.go#L29)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> s/pp/pper/
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6820)*
<!-- Sent from Reviewable.io -->

BramGruneir commented:
Review status: 0 of 2 files reviewed at latest revision, 23 unresolved discussions.

---

*[storage/reservation.go, line 50 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6820#-KIKBNr9OB3NHhxojL3W:-KIWQEQoXz344XOWhmeq:-1576657404) ([raw file](https://github.com/cockroachdb/cockroach/blob/939f4590198c9a2fafbb0917a7afd53f0ef0181f/storage/reservation.go#L50)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary>
> I'm going to convert it to a proper fifo queue. 
</details>
Done.  Went with a slice based queue instead of using list or rolling my own doubly linked list.  Using a slice is more in line with standard go idioms and should be efficient for a relatively small number of items.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6820)*
<!-- Sent from Reviewable.io -->

cuongdo commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary>
> Sorry about the spelling errors in comments.  I just switched to a new editor and it's lacking spellchecking in comments.  I'm going to see if I can do something about that.
</details>


---

Reviewed 2 of 2 files at r3.
Review status: all files reviewed at latest revision, 10 unresolved discussions.

---

*[storage/reservation.go, line 50 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6820#-KIKBNr9OB3NHhxojL3W:-KIY1Ch_yjKvPkrgJCUr:-1637013728) ([raw file](https://github.com/cockroachdb/cockroach/blob/939f4590198c9a2fafbb0917a7afd53f0ef0181f/storage/reservation.go#L50)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary>
> Done.  Went with a slice based queue instead of using list or rolling my own doubly linked list.  Using a slice is more in line with standard go idioms and should be efficient for a relatively small number of items.
</details>
I prefer the new implementation because of its simplicity. I suspect that performance won't be an issue, but we'll see how real-world performance pans out.

---

*[storage/reservation.go, line 158 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6820#-KIKEDSljYOFuSZvP7dU:-KIY0wV5_wyIDduu1H1G:686371639) ([raw file](https://github.com/cockroachdb/cockroach/blob/939f4590198c9a2fafbb0917a7afd53f0ef0181f/storage/reservation.go#L158)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary>
> I'm planning on adding that soon.  I'm not sure if that should fit within the reservation system or within the store.
</details>
Maybe add a `TODO` to make it more obvious to readers of code?

---

*[storage/reservation_test.go, line 48 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6820#-KIDcpq2Iv8ShZj6EJM2:-KIY1vMU2-e7T5c03-2_:1343548681) ([raw file](https://github.com/cockroachdb/cockroach/blob/800be2c7b142625528b383f255fc9782dfb4dd7c/storage/reservation_test.go#L48)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary>
> That seems like a bit much, there are only two actions.  
</details>
With the addition of the field names, this is no longer necessary.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6820#-:-KIY2UEm1qOodO6zf83r:1417088721)*
<!-- Sent from Reviewable.io -->

BramGruneir commented:
Review status: 1 of 2 files reviewed at latest revision, 10 unresolved discussions.

---

*[storage/reservation.go, line 50 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6820#-KIKBNr9OB3NHhxojL3W:-KIYMkvoRrHWWo0spvrg:-1611696665) ([raw file](https://github.com/cockroachdb/cockroach/blob/939f4590198c9a2fafbb0917a7afd53f0ef0181f/storage/reservation.go#L50)):*
<details><summary><i>Previously, cuongdo (Cuong Do) wrote…</i></summary>
> I prefer the new implementation because of its simplicity. I suspect that performance won't be an issue, but we'll see how real-world performance pans out.
</details>
The only reason to go with the linked-list implementation is that it removes the need for the `filled` attribute.  But I don't see this as being a real blocker.

---

*[storage/reservation.go, line 158 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6820#-KIKEDSljYOFuSZvP7dU:-KIYNLBK1bLA9Pkwo3J6:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/939f4590198c9a2fafbb0917a7afd53f0ef0181f/storage/reservation.go#L158)):*
<details><summary><i>Previously, cuongdo (Cuong Do) wrote…</i></summary>
> Maybe add a `TODO` to make it more obvious to readers of code?
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6820)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Review status: 1 of 2 files reviewed at latest revision, 10 unresolved discussions.

---

*[storage/reservation.go, line 50 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6820#-KIKBNr9OB3NHhxojL3W:-KIZdxMVYGydXZKM_VFW:-1434299435) ([raw file](https://github.com/cockroachdb/cockroach/blob/939f4590198c9a2fafbb0917a7afd53f0ef0181f/storage/reservation.go#L50)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary>
> The only reason to go with the linked-list implementation is that it removes the need for the `filled` attribute.  But I don't see this as being a real blocker.
</details>
`resByRangeID` contains only unfilled reservations, so you can use presence in that map instead of a separate `filled` attribute. This also removes several error code paths below.

---

*[storage/reservation.go, line 38 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6820#-KIZHqM7nVXJhu5lcf9b:-KIZHqM7nVXJhu5lcf9c:-457677649) ([raw file](https://github.com/cockroachdb/cockroach/blob/b2a9f024690a3a5323db305b528cba7822ffbe4d/storage/reservation.go#L38)):*
> ```Go
> }
> 
> // booking is an item in both the bookingPQ and the used in bookie's
> ```

s/P//

---

*[storage/reservation.go, line 42 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6820#-KIZemY33IbdVPReEqKb:-KIZemY33IbdVPReEqKc:1834650799) ([raw file](https://github.com/cockroachdb/cockroach/blob/b2a9f024690a3a5323db305b528cba7822ffbe4d/storage/reservation.go#L42)):*
> ```Go
> type booking struct {
> 	reservation reservation
> 	expireAt    roachpb.Timestamp // This is also the priority for the queue.
> ```

Remove the comment; no more priority queue.

---

*[storage/reservation.go, line 138 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6820#-KIZdGRhl1zmgk1zqVVX:-KIZdGRhl1zmgk1zqVVY:-1561737041) ([raw file](https://github.com/cockroachdb/cockroach/blob/b2a9f024690a3a5323db305b528cba7822ffbe4d/storage/reservation.go#L138)):*
> ```Go
> 		} else {
> 			if log.V(2) {
> 				log.Infof("there is pre-existing reservation for rangeID:%d, %v", res.rangeID, oldRes)
> ```

Log the source node/store IDs here, since that is what distinguishes the existing reservations that can be updated from the ones that cannot.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6820)*
<!-- Sent from Reviewable.io -->

BramGruneir commented:
Review status: 1 of 2 files reviewed at latest revision, 10 unresolved discussions.

---

*[storage/reservation.go, line 50 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6820#-KIKBNr9OB3NHhxojL3W:-KIZnK1sVnQxqzLhtu3p:2129167674) ([raw file](https://github.com/cockroachdb/cockroach/blob/939f4590198c9a2fafbb0917a7afd53f0ef0181f/storage/reservation.go#L50)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> `resByRangeID` contains only unfilled reservations, so you can use presence in that map instead of a separate `filled` attribute. This also removes several error code paths below.
</details>
Good call.  Done.

---

*[storage/reservation.go, line 38 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6820#-KIZHqM7nVXJhu5lcf9b:-KIZn82fkOKD4fcn00UZ:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/b2a9f024690a3a5323db305b528cba7822ffbe4d/storage/reservation.go#L38)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> s/P//
</details>
Done.

---

*[storage/reservation.go, line 42 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6820#-KIZemY33IbdVPReEqKb:-KIZnIJvc71IttdmZdNu:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/b2a9f024690a3a5323db305b528cba7822ffbe4d/storage/reservation.go#L42)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> Remove the comment; no more priority queue.
</details>
Done.

---

*[storage/reservation.go, line 138 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6820#-KIZdGRhl1zmgk1zqVVX:-KIZpCxuFKxPpPZCeJxk:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/b2a9f024690a3a5323db305b528cba7822ffbe4d/storage/reservation.go#L138)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> Log the source node/store IDs here, since that is what distinguishes the existing reservations that can be updated from the ones that cannot.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6820)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
LGTM

<details><summary><i>Previously, cuongdo (Cuong Do) wrote…</i></summary>
> :lgtm:
</details>


---

Review status: 1 of 2 files reviewed at latest revision, 10 unresolved discussions.

---

*[storage/reservation.go, line 94 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6820#-KIa5_ycRomrPCf_s2Hf:-KIa5_ydXkGya6hQFVcJ:1109543451) ([raw file](https://github.com/cockroachdb/cockroach/blob/732190d5e91114f52efa121f917a67ad382881d4/storage/reservation.go#L94)):*
> ```Go
> 	}
> 	b.mu.resByRangeID = make(map[roachpb.RangeID]*booking)
> 	b.mu.queue = []*booking{}
> ```

You can just leave `queue` with its zero value.

---

*[storage/reservation.go, line 161 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6820#-KIa6NLsQyO_jodKvFHM:-KIa6NLsQyO_jodKvFHN:1032840290) ([raw file](https://github.com/cockroachdb/cockroach/blob/732190d5e91114f52efa121f917a67ad382881d4/storage/reservation.go#L161)):*
> ```Go
> 
> 	// Lookup the reservation.
> 	res, ok := b.mu.resByRangeID[rangeID]
> ```

Don't check `resByRangeID` in both `fillBookingLocked` and its callers. Both `Fill` and `Reserve` got their reservation pointers from the map so there's no need to check it again and introduce additional error-handling logic for a case that can't happen. The only place where the identity check could fail is in the GC loop, so add a map lookup there.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6820#-:-KIa7OABcZw15ur8ewHk:676834233)*
<!-- Sent from Reviewable.io -->

BramGruneir commented:
Review status: 1 of 2 files reviewed at latest revision, 10 unresolved discussions.

---

*[storage/reservation.go, line 94 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6820#-KIa5_ycRomrPCf_s2Hf:-KIbzJwWG2W1MtdaMS2u:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/732190d5e91114f52efa121f917a67ad382881d4/storage/reservation.go#L94)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> You can just leave `queue` with its zero value.
</details>
Done.

---

*[storage/reservation.go, line 161 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6820#-KIa6NLsQyO_jodKvFHM:-KIc0QZOlpX54DWp2XUV:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/732190d5e91114f52efa121f917a67ad382881d4/storage/reservation.go#L161)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> Don't check `resByRangeID` in both `fillBookingLocked` and its callers. Both `Fill` and `Reserve` got their reservation pointers from the map so there's no need to check it again and introduce additional error-handling logic for a case that can't happen. The only place where the identity check could fail is in the GC loop, so add a map lookup there.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6820)*
<!-- Sent from Reviewable.io -->
